formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/schema/Generation.kt

```diff
@@ -0,0 +1,413 @@
+    index: Int,
+    builder: StringBuilder
+): List<SerialDescriptor> {
+    val fieldDescriptor = messageDescriptor.getElementDescriptor(index)
+    val nestedTypes: List<SerialDescriptor>
+    val typeName: String = when {
+        messageDescriptor.isSealedPolymorphic && index == 1 -> {
+            builder.appendLine("  // decoded as message with one of these types:")
+            nestedTypes = fieldDescriptor.elementDescriptors.toList()
+            nestedTypes.forEachIndexed { _, childDescriptor ->
+                builder.append("  //   message ").append(childDescriptor.messageOrEnumName)
+                    .append(", serial name '").append(removeLineBreaks(childDescriptor.serialName)).appendLine('\'')
+            }
+            fieldDescriptor.scalarTypeName()
+        }
+        fieldDescriptor.isProtobufScalar -> {
+            nestedTypes = emptyList()
+            fieldDescriptor.scalarTypeName(messageDescriptor.getElementAnnotations(index))
+        }
+        fieldDescriptor.isOpenPolymorphic -> {
+            nestedTypes = listOf(SyntheticPolymorphicDescriptor)
+            SyntheticPolymorphicDescriptor.serialName
+        }
+        else -> {
+            // enum or regular message
+            nestedTypes = listOf(fieldDescriptor)
+            fieldDescriptor.messageOrEnumName
+        }
+    }
+
+    if (fieldDescriptor.isNullable && messageDescriptor.isElementOptional(index)) {
+        builder.appendLine("  // WARNING: this field is nullable and has default value, it's impossible to unambiguously interpret an absence value.")
+        builder.appendLine("  //   For this field null value does not support and absence value denotes as default value. Default value is not present in the schema.")
+    }
+    if (!fieldDescriptor.isNullable && messageDescriptor.isElementOptional(index)) {
+        builder.appendLine("  // WARNING: an absence value is decoded as a default value that is not present in the schema")
+    }
+    val optional = fieldDescriptor.isNullable || messageDescriptor.isElementOptional(index)
+
+    builder.append("  ")
+        .append(if (optional) "optional " else "required ")
+        .append(typeName)
+
+    return nestedTypes
+}
+
+private fun generateMapType(
+    messageDescriptor: SerialDescriptor,
+    index: Int,
+    builder: StringBuilder
+): List<SerialDescriptor> {
+    val mapDescriptor = messageDescriptor.getElementDescriptor(index)
+    val fieldName = messageDescriptor.getElementName(index)
+    val valueDescriptor = mapDescriptor.getElementDescriptor(1).let {
+        if (it.isProtobufCollection) {
+            if (it.isNullable) {
+                builder.appendLine("  // WARNING: null value is not supported for nested collections")
+            }
+            val wrapperName = "${messageDescriptor.messageOrEnumName}_${fieldName}"
+            buildClassSerialDescriptor(wrapperName) {
+                element("value", it)
+            }
+        } else {
```

Comment:
[nit] you can add a `return` in the end of first branch, then `else` (what's more important -- its nesting) won't be necessary